### PR TITLE
Migrate RCCrossPlatformSupport, RCSystemInfo, RCLogLevel, RCTransaction

### LINKED ISSUE #541: Migrate RCSystemInfo

### LINKED ISSUE #556: Remove RCTransaction
And reference `Purchases.Transaction`
Also make Transaction API @objc-able

### LINKED ISSUE #607: Migrate RCCrossPlatformSupport.h

### PR DESCRIPTION
**Pull request by taquitos**
Resolves #607
Resolves #541
Resolves #556

### CODE REVIEW

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    @objc public static var isSandbox: Bool {
+        let url = Bundle.main.appStoreReceiptURL
+        guard let url = url else {
+            return false
+        }
+
+        let receiptURLString = url.path
+        return receiptURLString.contains("sandboxReceipt")
+    }
+    @objc public static func frameworkVersion() -> String { // TODO: automate the setting of this, if it hasn't been.
```
Comment (taquitos) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:55]: Has it been automated?

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        return UIApplication.shared.applicationState == UIApplication.State.background
+    }
+
+    @objc private var isAppExtension: Bool {
+        return Bundle.main.bundlePath.hasSuffix(".appex")
+    }
+
+    #endif
+
+    @available(*, deprecated, message: """
```
Comment (taquitos) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:138]: instead of crashing, we new pass an error back. This simplified testing, since M1 macs can test the particular crashes due to exception handling particularities.

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        Use (Swift) init(platformFlavor:, platformFlavorVersion:, finishTransactions:) throws
+        or (objc)  -initWithPlatformFlavor: platformFlavorVersion: finishTransactions: error:
+        """)
+    @objc public func initWith(platformFlavor: String?, platformFlavorVersion: String?, finishTransactions: Bool) -> Self {
```
Comment (taquitos) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:142]: let the fun begin.

**Comment by aboedo** [comment]
@taquitos is this ready for review? 
if so, the best way to indicate it is to request a review from the team

File: Purchases/Public/RCPurchases.m [hunk 4]
```diff
@@ -347,6 +344,17 @@ - (instancetype)initWithAppUserID:(nullable NSString *)appUserID
...
+#if TARGET_OS_IOS || TARGET_OS_TV
+#define APP_DID_BECOME_ACTIVE_NOTIFICATION_NAME UIApplicationDidBecomeActiveNotification
+#define APP_WILL_RESIGN_ACTIVE_NOTIFICATION_NAME UIApplicationWillResignActiveNotification
+#elif TARGET_OS_OSX
+#define APP_DID_BECOME_ACTIVE_NOTIFICATION_NAME NSApplicationDidBecomeActiveNotification
+#define APP_WILL_RESIGN_ACTIVE_NOTIFICATION_NAME NSApplicationWillResignActiveNotification
+#elif TARGET_OS_WATCH
+#define APP_DID_BECOME_ACTIVE_NOTIFICATION_NAME NSExtensionHostDidBecomeActiveNotification
+#define APP_WILL_RESIGN_ACTIVE_NOTIFICATION_NAME NSExtensionHostWillResignActiveNotification
+#endif
```
Comment (aboedo) [inline Purchases/Public/RCPurchases.m:356]: 🤔 
we'd still be able to define these, right? 

i.e.: would something like this work? 

```swift
import Foundation
#if os(iOS)
import UIKit
#elseif os(macOS)
import NSApplication
#elseif os(watchOS)
import UIKit
#endif

// todo: make internal
@objc(RCCrossPlatformSupport) public class CrossPlatformSupport: NSObject {
#if os(iOS)
    @objc public static let appDidBecomeActiveNotificationName = UIApplication.didBecomeActiveNotification
#elseif os(macOS)
    @objc public static let appDidBecomeActiveNotificationName = NSApplication.didBecomeActiveNotification
#elseif os(watchOS)
    @objc public static let appDidBecomeActiveNotificationName = NSExtensionHost.didBecomeActiveNotification
#endif
}
```

usage: 

```objective-c
[self.notificationCenter addObserver:self
                            selector:@selector(applicationDidBecomeActive:)
                                name:RCCrossPlatformSupport.appDidBecomeActiveNotificationName object:nil];

```


That way `RCPurchases` doesn't have to own the responsibility of knowing which notification is the correct one per platform.

Comment (aboedo) [reply Purchases/Public/RCPurchases.m:356]: it would work for some of the stuff in `RCCrossPlatformSupport`, but not all (for example, I don't think we'd be able to use this approach for importing the correct framework)

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 1 above)
Comment (aboedo) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:55]: it gets bumped through `fastlane bump`  and `fastlane bump_and_update_changelog` https://github.com/RevenueCat/purchases-ios/blob/main/fastlane/Fastfile#L50

Comment (aboedo) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:55]: updating the file path there should suffice

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+#if os(iOS) || os(tvOS) || targetEnvironment(macCatalyst)
+import UIKit
+#elseif os(watchOS)
+import UIKit
+import WatchKit
+#elseif os(macOS)
+import AppKit
+#endif
+
+public typealias RCSystemInfo = SystemInfo
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:20]: why is this needed?

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    #if targetEnvironment(macCatalyst)
+    static let PLATFORM_HEADER = "uikitformac"
+    #elseif os(iOS)
+    static let PLATFORM_HEADER = "iOS"
+    #elseif os(watchOS)
+    static let PLATFORM_HEADER = "watchOS"
+    #elseif os(tvOS)
+    static let PLATFORM_HEADER = "tvOS"
+    #elseif os(macOS)
+    static let PLATFORM_HEADER = "macOS"
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:33]: maybe we should take the opportunity to update to lowerCamelCase here to follow Swift conventions

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 2 above)
Comment (aboedo) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:138]: this class isn't public, so I think instead of deprecating we should just remove the old initializer and update the code to use the new one

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+    @objc public static func appVersion() -> String {
+        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? ""
+    }
+
+    @objc public static func buildVersion() -> String {
+        return Bundle.main.infoDictionary?["CFBundleVersion"] as? String ?? ""
+    }
+
+    @objc public static func platformHeader() -> String {
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:71]: these might make more sense as computed vars now or even lets, if they're available as soon as the app starts

File: PurchasesCoreSwift/Logging/Strings/ConfigureStrings.swift [hunk 8]
```diff
@@ -12,7 +12,7 @@ import Foundation
 @objc(RCConfigureStrings) public class ConfigureStrings: NSObject {
     @objc public var adsupport_not_imported: String { "AdSupport framework not imported. Attribution data incomplete." }
     @objc public var application_active: String { "applicationDidBecomeActive" }
-    @objc public var configuring_purchases_proxy_url_set: String {
+    @objc public class var configuring_purchases_proxy_url_set: String {
```
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Strings/ConfigureStrings.swift:15]: why is this needed?

File: Purchases/Public/RCPurchases.m (hunk 4 above)
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:356]: just read the rest of the code, it seems like this approach would fall pretty much in line with everything else done in SystemInfo

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 2 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:138]: Unfortunately, the class is public, the API tester can see it by `@import Purchases;` only.

File: PublicSDKAPITester/RCSystemInfoAPI.m [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+//  Created by Joshua Liebowitz on 6/29/21.
+//  Copyright © 2021 Purchases. All rights reserved.
+//
+
+@import Foundation;
+@import Purchases;
+
+#import "RCSystemInfoAPI.h"
+
+@implementation RCSystemInfoAPI
```
[the remark this answers is not in the document]
Comment (taquitos) [reply PublicSDKAPITester/RCSystemInfoAPI.m:14]: The great thing about the API tester is that it only sees public files since I have it setup to @import Purchases or @PurchasesCoreSwift
I build these files before I change the api, so if I'm understanding visibility correctly, what this project sees is what the public sees too.

File: Purchases/Attribution/RCAttributionFetcher.m [hunk 10]
```diff
@@ -48,7 +46,8 @@ - (instancetype)initWithDeviceCache:(RCDeviceCache *)deviceCache
 
 - (nullable NSString *)identifierForAdvertisers {
     if (@available(iOS 6.0, macOS 10.14, *)) {
-        Class <FakeASIdentifierManager> _Nullable asIdentifierManagerClass = [self.attributionFactory asIdentifierClass];
+        id klass = [self.attributionFactory asIdentifierClass];
+        Class <FakeASIdentifierManager> _Nullable asIdentifierManagerClass = klass;
```
[the remark this answers is not in the document]
Comment (taquitos) [reply Purchases/Attribution/RCAttributionFetcher.m:50]: No, this was to help in debugging, I'll revert this change.

File: PurchasesCoreSwift/Logging/Strings/ConfigureStrings.swift (hunk 8 above)
Comment (taquitos) [reply PurchasesCoreSwift/Logging/Strings/ConfigureStrings.swift:15]: It's not, I missed the usage of Strings.configure.
typically strings are held as a class var instead of an instance var, so this felt natural to me. I meant to go back and take a look at this, but I didn't. I'll revert this and go with `Strings.configure.`

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 5 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:20]: It no longer is. Since I was migrating so many files, the reference to this class needed RCSystemInfo for a minute.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 6 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:33]: Sounds good to me

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 1 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:55]: Awesome.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 7 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:71]: Yeah, that was my first pass, turns out this is public API and it uses `()`. We could make vars and deprecate these funcs, though.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 1 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:53]: Updated in latest commit

File: Purchases.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -17,7 +17,6 @@
 		2D22679125F2D9AD00E6950C /* PurchasesTests-Bridging-Header.h in Headers */ = {isa = PBXBuildFile; fileRef = 35262A291F7D783F00C04F2C /* PurchasesTests-Bridging-Header.h */; };
 		2D390255259CF46000DB19C0 /* MockPaymentDiscount.swift in Sources */ = {isa = PBXBuildFile; fileRef = 2D390254259CF46000DB19C0 /* MockPaymentDiscount.swift */; };
 		2D3E29CE25E8009000456FA8 /* MockAttributionTypeFactory.swift in Sources */ = {isa = PBXBuildFile; fileRef = 37E35DA3E083FE37DAF15954 /* MockAttributionTypeFactory.swift */; };
-		2D4C02E426697E490038F877 /* RCLogLevel.h in Headers */ = {isa = PBXBuildFile; fileRef = 2D4C02E326697E490038F877 /* RCLogLevel.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
[the remark this answers is not in the document]
Comment (taquitos) [reply Purchases.xcodeproj/project.pbxproj:20]: It's probably best to take a look at the files deleted/files added for a sense of what's changed. I don't pay much attention to the changes in the project file since they are often red herrings. I find moving anything, or adding anything can set off a chain of changes that is completely inconsequential. If there's something specific about the project file that doesn't make sense, I'm happy to investigate, though!

[the remark this answers is not in the document]
Comment (taquitos) [reply Purchases.xcodeproj/project.pbxproj:20]: That being said, if I update build settings, I'll try to point that out better

File: Purchases/Public/RCPurchases.m (hunk 4 above)
Comment (taquitos) [reply Purchases/Public/RCPurchases.m:356]: HAHahahahahahahahaha you'd think, and I thought so to.
BUT that's not the case at all. I didn't dig into too much and opted to move the `#if` out of the file into the places where they were used since they only applied to 1 location for each `#define.`
Turns out there are subtle differences when you do defines in objc vs Swift and crossing the objc/swift barrier. 
Once we get to 100% swift, we can go back and clean up any weird stuff like this. Right now, our focus is move this over to swift while maintaining functionality during the whole process. Once we're done, there will be plenty of opportunities to make it more swifty and cleaner 😄

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 6 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift]: note: `platformHeader` is already used as a property, so I used `platformHeaderConstant`

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 2 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:138]: Discussed with @aboedo and did some more digging, it isn't public 🎉, updating

File: PublicSDKAPITester/RCSystemInfoAPI.m (hunk 9 above)
[the remark this answers is not in the document]
Comment (taquitos) [reply PublicSDKAPITester/RCSystemInfoAPI.m:14]: Will remove this.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 7 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:71]: Since it's not public, I'll switch them back to vars.

File: Purchases/Public/RCPurchases.m (hunk 4 above)
Comment (aboedo) [reply Purchases/Public/RCPurchases.m:366]: yeah, the `#define` is not a thing in Swift, and `#if` checks are certainly less flexible, so it wouldn't work stuff like `UI_DEVICE_AVAILABLE`, but why wouldn't the approach above work for `appDidBecomeActiveNotificationName`? Isn't it the same thing we're doing in `platformHeaderConstant`? 

I think the distinction here is that we can't use it to define anything that's pre-processor, but we should be able to use it for regular variables, like in this case, a string

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        return Bundle.main.bundlePath.hasSuffix(".appex")
+    }
+    #endif
+
+    @objc required public init(platformFlavor: String?, platformFlavorVersion: String?, finishTransactions: Bool) throws {
+        if let platformFlavor = platformFlavor {
+            self.platformFlavor = platformFlavor
+        } else {
+            self.platformFlavor = "native"
+        }
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:139]: these might be clearer as 
```swift
self.platformFlavor = platformFlavor ?? "native"
```

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+        if let platformFlavorVersion = platformFlavorVersion {
+            self.platformFlavorVersion = platformFlavorVersion
+        } else {
+            self.platformFlavorVersion = nil
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:144]: this if / else isn't really necessary, right? i.e.: just assigning to the optional would be the same?

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+    #elseif os(watchOS)
+        return  WKExtension.shared().applicationState == WKApplicationState.background
+    #endif
+    }
+
+    #if os(iOS)
+    // iOS App extensions can't access UIApplication.sharedApplication, and will fail to compile if any calls to
+    // it are made. There are no pre-processor macros available to check if the code is running in an app extension,
+    // so we check if we're running in an app extension at runtime, and if not, we use KVC to call sharedApplication.
+    @objc private var isApplicationBackgroundedIOS: Bool {
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:121]: if we have stuff that's private, I don't think we should have them be `objc` - we're not getting any value out of it, but we're adding restrictions for the types that can be used, in a codebase that will soon be swift-only

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+                                .replacingOccurrences(of: "%@", with: privateProxyURLString))
+            }
+        }
+    }
+
+    static var defaultServerHostURL: URL {
+        return URL(string: defaultServerHostName)!
+    }
+
+    @objc public static var serverHostURL: URL {
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:101]: I think we should try to keep the public stuff at the top, so the file is easier to read for anyone who's not looking to modify it

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+        if (platformFlavor == nil && platformFlavorVersion != nil) ||
+            (platformFlavor != nil && platformFlavorVersion == nil) {
+            Logger.error("RCSystemInfo initialized with non-matching platform flavor and platform flavor versions!")
+            throw SystemInfoError.invalidInitializationData
+        }
+
+        self.finishTransactions = finishTransactions
+    }
+
+    @objc open func isApplicationBackgrounded(completion: @escaping (Bool) -> Void) {
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:156]: why is this one `open`, vs the rest of the methods declared either `public`, `internal` or `private`?

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+        DispatchQueue.main.async {
+            completion(self.isApplicationBackgrounded)
+        }
+    }
+
+    open func isOperatingSystem(atLeastVersion version: OperatingSystemVersion) -> Bool {
+        return isOperatingSystemAtLeastVersion(version)
+    }
+
+    @objc public func isOperatingSystemAtLeastVersion(_ version: OperatingSystemVersion) -> Bool {
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:166]: why is this method defined twice? wouldn't just adding `@objc public` to the first version provide the same obj-c API?

File: PurchasesTests/Misc/SystemInfoTests.swift [hunk 18]
```diff
@@ -4,71 +4,66 @@ import Nimble
...
+        expect(SystemInfo.proxyURL).to(beNil())
 
         let url = URL(string: "https://my_url")
-        RCSystemInfo.setProxyURL(url)
+        SystemInfo.proxyURL = url
 
-        expect(RCSystemInfo.serverHostURL()) == url
-        expect(RCSystemInfo.proxyURL()) == url
+        expect(SystemInfo.serverHostURL) == url
+        expect(SystemInfo.proxyURL) == url
```
Comment (aboedo) [inline PurchasesTests/Misc/SystemInfoTests.swift:16]: 👏

File: PurchasesTests/Misc/SystemInfoTests.swift [hunk 19]
```diff
@@ -4,71 +4,66 @@ import Nimble
...
-            _ = RCSystemInfo(platformFlavor: nil, platformFlavorVersion: "b", finishTransactions: true)
-        }
-        expectToThrowException(.parameterAssert) {
-            _ = RCSystemInfo(platformFlavor: "a", platformFlavorVersion: nil, finishTransactions: true)
-        }
-            
-        expectToNotThrowException {
-            _ = RCSystemInfo(platformFlavor: nil, platformFlavorVersion: nil, finishTransactions: true)
-        }
+        expect(try SystemInfo(platformFlavor: "a", platformFlavorVersion: "b", finishTransactions: true)).toNot(throwError(SystemInfo.SystemInfoError.invalidInitializationData))
```
Comment (aboedo) [inline PurchasesTests/Misc/SystemInfoTests.swift:45]: awesome! 
although I honestly feel like breaking line here might make it easier to read

File: fastlane/Fastfile [hunk 20]
```diff
@@ -58,7 +58,7 @@ platform :ios do
     files_to_update = [
       '../PurchasesCoreSwift.podspec',
       '../Purchases.podspec',
-      '../Purchases/Misc/RCSystemInfo.m',
+      '../PurchasesCoreSwift/Misc/SystemInfo.swift',
```
Comment (aboedo) [inline fastlane/Fastfile:61]: 👌

File: Purchases/Public/RCPurchases.m (hunk 4 above)
Comment (taquitos) [reply Purchases/Public/RCPurchases.m:366]: Yeah, that would work, but since we're also migrating that file, I felt that putting those defines in the only file that uses it was the right course of action for now. I'm trying to change as little as possible. When we migrate that file, we'll want to switch to what you're suggesting. At this point, I should note, the constants file `RCCrossPlatformSupport` (`CrossPlatformSupport`) has been removed in favor of moving the `#define`s and `#if`s to the location where they are used. We can definitely revisit it after migration, it was just much easier for me to deal with the dependencies this way.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 12 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:139]: 🤦‍♂️

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 13 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:144]: Definitely, yeah. This was part of my debugging. I was trying to make sure I wasn't somehow missing something so I made if/else super explicit for a minute. Then forgot to go back and revise.
🤦‍♂️ again.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 15 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:101]: I agree, with a small caveat that I group private vars that back public vars right above the public get/set declaration.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 16 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:156]: Our tests subclass this and override this method.

File: PurchasesCoreSwift/Misc/SystemInfo.swift (hunk 17 above)
Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift]: That was when I thought the API was public and needed backwards compatibility with the swift name. Removing!

File: PurchasesTests/Misc/SystemInfoTests.swift (hunk 18 above)
Comment (taquitos) [reply PurchasesTests/Misc/SystemInfoTests.swift:16]: 🕺

File: PurchasesTests/Misc/SystemInfoTests.swift (hunk 19 above)
Comment (taquitos) [reply PurchasesTests/Misc/SystemInfoTests.swift:45]: Totally agree, updating.

**Comment by taquitos** [comment]
Merging, but feel free to add any further comments, I think this has gotten to a good state.

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+            return nil
+        #endif
+    }
+
+    private static var defaultServerHostURL: URL {
+        return URL(string: defaultServerHostName)!
+    }
+
+    @objc public static var serverHostURL: URL {
+        return self.proxyURL ?? Self.defaultServerHostURL
```
Comment (beylmk) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:89]: is the second `self` meant to be capitalized?

Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:89]: Ooooo since this is a type variable, I'm actually surprised `self.proxyURL` works since we're accessing a variable on the type, which should be done with `Self` 
I'll fix the lowercase `self`.

Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:89]: https://github.com/RevenueCat/purchases-ios/pull/615 🎉

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+        get {
+            return privateProxyURL
+        }
+        set {
+            privateProxyURL = newValue
+            if let privateProxyURLString = newValue?.absoluteString {
+                Logger.info(Strings.configure.configuring_purchases_proxy_url_set
+                                .replacingOccurrences(of: "%@", with: privateProxyURLString))
+            }
+        }
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:103]: why is this private `proxyURL` needed?
would a `didSet` solve it? 
i.e.: 

```swift
@objc public static var proxyURL: URL? { 
    didSet { 
        if let proxyURL = proxyURL { 
             Logger.info(Strings.configure.configuring_purchases_proxy_url_set
                                .replacingOccurrences(of: "%@",with: privateProxyURLString))
        }
    }
}
```

Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:103]: Well, uh, yeah. Of course it would 🤦‍♂️ nice catch.

Comment (taquitos) [reply PurchasesCoreSwift/Misc/SystemInfo.swift:103]: https://github.com/RevenueCat/purchases-ios/pull/620

File: PurchasesCoreSwift/Misc/SystemInfo.swift [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+    #if os(iOS)
+    // iOS App extensions can't access UIApplication.sharedApplication, and will fail to compile if any calls to
+    // it are made. There are no pre-processor macros available to check if the code is running in an app extension,
+    // so we check if we're running in an app extension at runtime, and if not, we use KVC to call sharedApplication.
+    private var isApplicationBackgroundedIOS: Bool {
+        if self.isAppExtension {
+            return true
+        }
+
+        return UIApplication.shared.applicationState == UIApplication.State.background
```
Comment (aboedo) [inline PurchasesCoreSwift/Misc/SystemInfo.swift:150]: this line breaks compilation for app extensions. `shared` isn't available for them, and it will fail at compilation time, which is why we were doing it this way before: https://github.com/RevenueCat/purchases-ios/blob/main/Purchases/Misc/RCSystemInfo.m#L139
